ip: Merge ip stack from current

cathay4t opened the pull request:
Since the `InterfaceIpv4` and `InterfaceIpv6` are using `enabled: false`
by default, this prevent user from applying this desire state:

```yml
---
interfaces:
  - name: eth1
    type: ethernet
    state: up
    mtu: 1500
    ipv4:
      dhcp: true
    ipv6:
      dhcp: true
      autoconf: true
```

We should merge ip stack as we did for python API.

Implementation detail:
 * Introducing private struct `InterfaceIp` using serde build-in
   Deserializer and Serializer.
 * Implement `Deserialize` for `InterfaceIpv4` and `InterfaceIpv6`
   preserving the `prop_list` which will used to determined whether user
   desired `enabled` property or not.
 * At `Interfaces` struct, we merge ip stack `enabled`, `dhcp`,
   `autoconf` properties before applying and verifying.

Design choice:
 * Why not just convert `enabled: bool` to `enabled: Option<bool>`?
    1. This break rust API of nmstate.
    2. A lot lines need to changed and handle `enabled: None` case which
       might be interpreted in differently in the future or by other
       developer. The best approach should be do special merge action
       and follow up code don't need to worry about `enabled: None`
       use case.

Integration test case included.

ffmancera commented on rust/src/lib/ip.rs:
```diff
@@ -667,3 +797,87 @@ pub(crate) fn validate_wait_ip(
+        }
+        if let Some(cur_ip_conf) = cur_iface.base_iface().ipv4.as_ref() {
+            if let Some(ip_conf) = iface.base_iface_mut().ipv4.as_mut() {
+                println!("HAHA ip_conf {:?}", ip_conf);
+                println!("HAHA cur_ip_conf {:?}", cur_ip_conf);
```
These are leftovers from debuggin, please remove them

ffmancera commented on rust/src/lib/ip.rs:
```diff
@@ -667,3 +797,87 @@ pub(crate) fn validate_wait_ip(
+                println!("HAHA ip_conf {:?}", ip_conf);
+                println!("HAHA cur_ip_conf {:?}", cur_ip_conf);
+                if !ip_conf.prop_list.contains(&"enabled") {
+                    println!("HAHA");
```
same here

ffmancera commented on rust/src/lib/ip.rs:
```diff
@@ -667,3 +797,87 @@ pub(crate) fn validate_wait_ip(
+                if ip_conf.dhcp.is_none() && ip_conf.enabled {
+                    ip_conf.dhcp = cur_ip_conf.dhcp;
+                }
+                println!("HAHA new ip_conf {:?}", ip_conf);
```
same here

ffmancera reviewed (changes requested):
Thank you! It looks good to me, there are some leftovers.. Thanks!

cathay4t committed:
ip: Merge ip stack from current

Since the `InterfaceIpv4` and `InterfaceIpv6` are using `enabled: false`
by default, this prevent user from applying this desire state:

```yml
---
interfaces:
  - name: eth1
    type: ethernet
    state: up
    mtu: 1500
    ipv4:
      dhcp: true
    ipv6:
      dhcp: true
      autoconf: true
```

We should merge ip stack as we did for python API.

Implementation detail:
 * Introducing private struct `InterfaceIp` using serde build-in
   Deserializer and Serializer.
 * Implement `Deserialize` for `InterfaceIpv4` and `InterfaceIpv6`
   preserving the `prop_list` which will used to determined whether user
   desired `enabled` property or not.
 * At `Interfaces` struct, we merge ip stack `enabled`, `dhcp`,
   `autoconf` properties before applying and verifying.

Design choice:
 * Why not just convert `enabled: bool` to `enabled: Option<bool>`?
    1. This break rust API of nmstate.
    2. A lot lines need to changed and handle `enabled: None` case which
       might be interpreted in differently in the future or by other
       developer. The best approach should be do special merge action
       and follow up code don't need to worry about `enabled: None`
       use case.

Integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>
